code/game/objects/items/storage/storage.dm

```diff
@@ -71,24 +110,31 @@
 				usr.dropItemToGround(src)
 				usr.put_in_l_hand(src)
 
-/obj/item/storage/proc/return_inv()
-
-	var/list/L = list(  )
-
-	L += src.contents
-
-	for(var/obj/item/storage/S in src)
-		L += S.return_inv()
-	for(var/obj/item/gift/G in src)
-		L += G.gift
-		if (istype(G.gift, /obj/item/storage))
-			L += G.gift:return_inv()
-	return L
+///Recursively gets all contents and sub-contents of this storage item
+/obj/item/storage/proc/return_inv(recurse = TRUE)
+	. = list()
+	. += contents
 
-/obj/item/storage/proc/show_to(mob/user as mob)
+	for(var/content in contents)
+		if(isstorage(content))
+			var/obj/item/storage/substorage = content
+			. += substorage.return_inv(recurse)
+			continue
+		if(istype(content, /obj/item/gift))
+			var/obj/item/gift/subgift = content
+			. += subgift
+			if(isstorage(subgift.gift))
+				var/obj/item/storage/substorage = subgift.gift
+				. += substorage.return_inv(recurse)
+
+///Adds and opens the view of this storage for the given mob
+/obj/item/storage/proc/show_to(mob/user)
 	if(user.s_active != src)
-		for(var/obj/item/I in src)
-			if(I.on_found(user))
+		for(var/content in contents)
+			if(!isitem(content))
+				continue
+			var/obj/item/item = content
```

Comment:
if there is supposedly NO WAY something that aint an item being in contents just add AS, if something that should not be and happens to be fix that instead.
```suggestion
		for(var/obj/item/content in contents)
```